File: src/lib/firebase/repositories/productRepository.ts
import { 
  collection,
  doc,
  getDocs,
  getDoc,
  addDoc,
  updateDoc,
  deleteDoc,
  query,
  where,
  orderBy,
  onSnapshot,
  Timestamp,
  Unsubscribe
} from 'firebase/firestore';
import { db } from '../config';
import { Product } from '@/types/models';

/**
 * Firebase Product Repository - Replicates Android product queries
 * Handles products with same patterns as Android app
 */
class FirebaseProductRepository {
  private static instance: FirebaseProductRepository;
  
  private constructor() {}
  
  static getInstance(): FirebaseProductRepository {
    if (!FirebaseProductRepository.instance) {
      FirebaseProductRepository.instance = new FirebaseProductRepository();
    }
    return FirebaseProductRepository.instance;
  }
  
  /**
   * Get all products for a specific shop (matches Android pattern)
   * Uses combined query to satisfy Firestore security rules
   */
  async getProductsByShop(shopId: string): Promise<Product[]> {
    try {
      console.log(`🔥 Fetching products for shop ${shopId}`);
      
      // Combined query to satisfy Firestore security rules (only active products can be read)
      const productsQuery = query(
        collection(db, 'products'),
        where('shopId', '==', shopId),
        where('active', '==', true)
      );
      
      const snapshot = await getDocs(productsQuery);
      console.log(`🔥 Fetched ${snapshot.docs.length} active product documents for shop`);
      
      const products: Product[] = [];
      snapshot.docs.forEach((document) => {
        try {
          const data = document.data();
          const product: Product = {
            id: document.id,
            ...data,
            createdAt: data.createdAt || Date.now(),
            updatedAt: data.updatedAt || Date.now(),
          } as Product;
          
          console.log(`✅ Successfully parsed active product for shop: ${product.name}`);
          products.push(product);
        } catch (e) {
          console.error(`❌ Failed to parse product document ${document.id}`, e);
          // Skip invalid documents like Android does
        }
      });
      
      // Sort in code instead of query (like Android does)
      return products.sort((a, b) => b.createdAt - a.createdAt);
    } catch (error) {
      console.error('❌ Error fetching shop products:', error);
      return [];
    }
  }
  
  /**
   * Get all active products (for marketplace view)
   */
  async getActiveProducts(): Promise<Product[]> {
    try {
      console.log('🔥 Fetching all active products');
      
      // Simplified query without orderBy to avoid index requirements
      const productsQuery = query(
        collection(db, 'products'),
        where('active', '==', true)
      );
      
      const snapshot = await getDocs(productsQuery);
      console.log(`🔥 Fetched ${snapshot.docs.length} active product documents`);
      
      const products: Product[] = [];
      snapshot.docs.forEach((document) => {
        try {
          const data = document.data();
          const product: Product = {
            id: document.id,
            ...data,
            createdAt: data.createdAt || Date.now(),
            updatedAt: data.updatedAt || Date.now(),
          } as Product;
          
          console.log(`✅ Successfully parsed active product: ${product.name}`);
          products.push(product);
        } catch (e) {
          console.error(`❌ Failed to parse product document ${document.id}`, e);
        }
      });
      
      console.log(`✅ Successfully parsed ${products.length} active products`);
      return products;
    } catch (error) {
      console.error('❌ Error fetching active products:', error);
      return [];
    }
  }
  
  /**
   * Get a product by its unique ID
   */
  async getProductById(id: string): Promise<Product | null> {
    try {
      const docRef = doc(db, 'products', id);
      const document = await getDoc(docRef);
      
      if (!document.exists()) {
        return null;
      }
      
      const data = document.data();
      const product: Product = {
        id: document.id,
        ...data,
        createdAt: data.createdAt || Date.now(),
        updatedAt: data.updatedAt || Date.now(),
      } as Product;
      
      return product;
    } catch (error) {
      console.error(`❌ Error fetching product ${id}:`, error);
      return null;
    }
  }
  
  /**
   * Get products by owner ID (for business dashboard)
   */
  async getProductsByOwner(ownerId: string): Promise<Product[]> {
    try {
      const productsQuery = query(
        collection(db, 'products'),
        where('ownerId', '==', ownerId),
        orderBy('createdAt', 'desc')
      );
      
      const snapshot = await getDocs(productsQuery);
      
      const products: Product[] = [];
      snapshot.docs.forEach((document) => {
        try {
          const data = document.data();
          const product: Product = {
            id: document.id,
            ...data,
            createdAt: data.createdAt || Date.now(),
            updatedAt: data.updatedAt || Date.now(),
          } as Product;
          
          products.push(product);
        } catch (e) {
          console.error(`❌ Failed to parse product document ${document.id}`, e);
        }
      });
      
      return products;
    } catch (error) {
      console.error(`❌ Error fetching products for owner ${ownerId}:`, error);
      return [];
    }
  }
  
  /**
   * Add a new product
   */
  async addProduct(product: Omit<Product, 'id'>): Promise<string | null> {
    try {
      const productData = {
        ...product,
        createdAt: Date.now(),
        updatedAt: Date.now(),
      };
      
      const docRef = await addDoc(collection(db, 'products'), productData);
      return docRef.id;
    } catch (error) {
      console.error('❌ Error adding product:', error);
      return null;
    }
  }
  
  /**
   * Update an existing product
   */
  async updateProduct(product: Product): Promise<boolean> {
    try {
      const docRef = doc(db, 'products', product.id);
      const { id, createdAt, ...updateData } = product;
      
      await updateDoc(docRef, {
        ...updateData,
        updatedAt: Date.now(),
      });
      
      return true;
    } catch (error) {
      console.error(`❌ Error updating product ${product.id}:`, error);
      return false;
    }
  }
  
  /**
   * Delete a product
   */
  async deleteProduct(id: string): Promise<boolean> {
    try {
      const docRef = doc(db, 'products', id);
      await deleteDoc(docRef);
      return true;
    } catch (error) {
      console.error(`❌ Error deleting product ${id}:`, error);
      return false;
    }
  }
  
  /**
   * Search products by name or description
   */
  async searchProducts(searchTerm: string, limit: number = 20): Promise<Product[]> {
    try {
      // Note: Firestore doesn't have full-text search, so this is a simple implementation
      // For production, consider using Algolia or similar
      const productsQuery = query(
        collection(db, 'products'),
        where('isActive', '==', true),
        orderBy('name')
      );
      
      const snapshot = await getDocs(productsQuery);
      
      const products: Product[] = [];
      const searchLower = searchTerm.toLowerCase();
      
      snapshot.docs.forEach((document) => {
        try {
          const data = document.data();
          const product: Product = {
            id: document.id,
            ...data,
            createdAt: data.createdAt || Date.now(),
            updatedAt: data.updatedAt || Date.now(),
          } as Product;
          
          // Simple text matching
          if (
            product.name.toLowerCase().includes(searchLower) ||
            product.description.toLowerCase().includes(searchLower) ||
            product.tags.some(tag => tag.toLowerCase().includes(searchLower))
          ) {
            products.push(product);
          }
        } catch (e) {
          console.error(`❌ Failed to parse product document ${document.id}`, e);
        }
      });
      
      return products.slice(0, limit);
    } catch (error) {
      console.error('❌ Error searching products:', error);
      return [];
    }
  }
}

// Export singleton instance
export const productRepository = FirebaseProductRepository.getInstance();

File: androidAPP/MyHustle/app/src/main/java/com/blueclipse/myhustle/data/repository/ProductRepository.kt
package com.blueclipse.myhustle.data.repository

import com.blueclipse.myhustle.data.model.Product
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.tasks.await

/**
 * Repository for managing product data in Firebase
 */
class ProductRepository private constructor() {
    private val firestore = FirebaseFirestore.getInstance()
    private val productsCollection = firestore.collection("products")
    
    private val _products = MutableStateFlow<List<Product>>(emptyList())
    val products: StateFlow<List<Product>> = _products.asStateFlow()
    
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    
    companion object {
        @Volatile
        private var INSTANCE: ProductRepository? = null
        
        val instance: ProductRepository
            get() = INSTANCE ?: synchronized(this) {
                INSTANCE ?: ProductRepository().also { INSTANCE = it }
            }
    }
    
    /**
     * Create a new product
     */
    suspend fun createProduct(product: Product): Result<String> {
        return try {
            _isLoading.value = true
            
            val documentRef = productsCollection.add(product).await()
            val productId = documentRef.id
            
            // Update the product with its ID
            documentRef.update("id", productId).await()
            
            Result.success(productId)
        } catch (e: Exception) {
            println("Error creating product: ${e.message}")
            e.printStackTrace()
            Result.failure(e)
        } finally {
            _isLoading.value = false
        }
    }
    
    /**
     * Get products for a specific shop
     */
    suspend fun getProductsForShop(shopId: String): Result<List<Product>> {
        return try {
            _isLoading.value = true
            val snapshot = productsCollection
                .whereEqualTo("shopId", shopId)
                .get()
                .await()
            
            val products = snapshot.documents.mapNotNull { document ->
                try {
                    val product = document.toObject(Product::class.java)?.copy(id = document.id)
                    // Filter active products in code instead of query to avoid index requirement
                    if (product?.isActive == true) product else null
                } catch (e: Exception) {
                    null
                }
            }.sortedByDescending { it.createdAt } // Sort in code instead of query
            
            _products.value = products
            Result.success(products)
        } catch (e: Exception) {
            Result.failure(e)
        } finally {
            _isLoading.value = false
        }
    }
    
    /**
     * Get all active products
     */
    suspend fun getAllProducts(): Result<List<Product>> {
        return try {
            _isLoading.value = true
            val snapshot = productsCollection
                .whereEqualTo("isActive", true)
                .orderBy("createdAt", Query.Direction.DESCENDING)
                .get()
                .await()
            
            val products = snapshot.documents.mapNotNull { document ->
                try {
                    document.toObject(Product::class.java)?.copy(id = document.id)
                } catch (e: Exception) {
                    null
                }
            }
            
            Result.success(products)
        } catch (e: Exception) {
            Result.failure(e)
        } finally {
            _isLoading.value = false
        }
    }
    
    /**
     * Get product by ID
     */
    suspend fun getProductById(productId: String): Result<Product?> {
        return try {
            val document = productsCollection.document(productId).get().await()
            if (document.exists()) {
                val product = document.toObject(Product::class.java)?.copy(id = document.id)
                Result.success(product)
            } else {
                Result.success(null)
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    /**
     * Update product
     */
    suspend fun updateProduct(product: Product): Result<Unit> {
        return try {
            val updateData = product.copy(updatedAt = System.currentTimeMillis())
            productsCollection.document(product.id).set(updateData).await()
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    /**
     * Delete product
     */
    suspend fun deleteProduct(productId: String): Result<Unit> {
        return try {
            productsCollection.document(productId).delete().await()
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    /**
     * Update product stock
     */
    suspend fun updateStock(productId: String, newQuantity: Int): Result<Unit> {
        return try {
            val updateData = mapOf(
                "stockQuantity" to newQuantity,
                "inStock" to (newQuantity > 0),
                "updatedAt" to System.currentTimeMillis()
            )
            
            productsCollection.document(productId).update(updateData).await()
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}
